sklearn/naive_bayes.py

```diff
@@ -951,3 +964,182 @@ def _joint_log_likelihood(self, X):
         jll += self.class_log_prior_ + neg_prob.sum(axis=1)
 
         return jll
+
+
+class CategoricalNB(BaseDiscreteNB):
+    """
+    Naive Bayes classifier for categorical features
+
+    The categorical Naive Bayes classifier is suitable for classification with
+    discrete features, that are categorically distributed. Each feature is sampled
+    by its own categorical distribution.
+
+    Read more in the :ref:`User Guide <categorical_naive_bayes>`.
+
+    Parameters
+    ----------
+    alpha : float, optional (default=1.0)
+        Additive (Laplace/Lidstone) smoothing parameter
+        (0 for no smoothing).
+
+    fit_prior : boolean, optional (default=True)
+        Whether to learn class prior probabilities or not.
+        If false, a uniform prior will be used.
+
+    class_prior : array-like, size (n_classes,), optional (default=None)
+        Prior probabilities of the classes. If specified the priors are not
+        adjusted according to the data.
+
+    Attributes
+    ----------
+    class_log_prior_ : array, shape (n_classes, )
+        Smoothed empirical log probability for each class.
+
+    feature_log_prob_ : array, shape (n_classes, n_features)
+        Empirical log probability of features
+        given a class, ``P(x_i|y)``.
+
+    class_count_ : array, shape (n_classes,)
+        Number of samples encountered for each class during fitting. This
+        value is weighted by the sample weight when provided.
+
+    feature_count_ : dict of arrays, len n_features
```

Comment:
Should this rather be a 2d array? Perhaps call it `category_count_`?

Reply:
I changed the wording to `category_count_`. I, too, think this is easier to understand.
However `category_count_` can't easily be 2D. If it would be 2D with shape (n_samples, n_features) the information of the categories for each feature would be missing. However, it is not straightforward to extend it to 3D, because generally speaking each feature has a different categorical distribution with a different number of categories.
It could be possible though, to use a jagged array here with variable length 3rd dimension. Is that what you were thinking of?

Reply:
Or how about a list of arrays? This would at least make the mapping of the column to the appropriate element in `category_count_` easier.